model/fn-execution/src/main/proto/beam_fn_api.proto

```diff
@@ -257,6 +262,122 @@ message ProcessBundleProgressRequest {
+    Metric metric = 4;
+  }
+
+  // A set of key+value labels which define the scope of the metric.
+  // Either a well defined entity id for the keys:
+  // “transform”, “pcollection”, “windowing_strategy”,
+  // “coder”, “environment” or any arbitrary label
+  // set by a custom metric or user metric.
+  // A monitoring system is expected to be able to aggregate the metric together
+  // for all updates having the same URN and labels.
+  // Some systems such as Stackdriver will be able to aggregate the metric
+  // using a subset of the provided labels
+  map<string, string> labels = 5;
+}
+
+message Metric {
+  // (Required) The data for this metric.
+  oneof data {
+    CounterData counter_data = 1;
+    DistributionData distribution_data = 2;
+    Extrema extrema_data = 3;
+  }
+}
+
+// Data associated with a Counter or Gauge metric.
+// This is designed to be compatible with metric collection
+// systems such as DropWizard.
+message CounterData {
+   oneof value {
+     int64 int64_value = 1;
+     string string_value = 2;
+     double double_value = 3;
+   }
+}
+
+// Extrema messages are used for calculating
+// Top-N/Bottom-N metrics.
+message Extrema {
+  // Only one of the two should be specified.
+  // Note: oneof is not allowed on repeated fields.
+  repeated int64 int_values = 1;
+  repeated double double_values = 2;
+}
+
+// Data associated with a distribution metric.
+// This is based off of the current DistributionData metric
+// This is not a stackdriver or dropwizard compatible
+// style of distribution metric.
+message DistributionData {
+  oneof distribution {
+    IntDistributionData int_double_distribution = 1;
+    DoubleDistributionData double_distribution_data =2;
+  }
+}
+
+message IntDistributionData {
+  int64 count = 1;
+  int64 sum = 2;
+  int64 min = 3;
+  int64 max = 4;
+}
+
+message DoubleDistributionData {
+  int64 count = 1;
+  double sum = 2;
+  double min = 3;
+  double max = 4;
+}
+
+// General MonitoredState information which contains
+// structured information which does not fit into a typical
+// metric format. For example, a table of important files
+// and metadata which an I/O source is reading.
+// Note: Since MonitoredState is designed to be
+// customizable, and allow engines to aggregate these
+// metrics in custom ways, there is no enforced aggregation function
+// The main expectation is that an SDK should report
+// all relevant MonitoredState for its PTransform, essentially
+// ‘overwriting’ what it previously sent. This means a RunnerHarness
+// should treat new MonitoredState updates as overwriting old ones. 
+message MonitoringTableData {
+  message MonitoringColumnValue {
+    oneof value {
+      int64 int64_value = 1;
+      double double_value = 2;
+      string string_value = 3;
+      google.protobuf.Timestamp timestamp = 4;
+    }
+  }
+
+  message MonitoringRow {
+    repeated MonitoringColumnValue values = 1;
+  }
+
+ repeated string column_names = 1;
```

Comment:
Mention that the number of column_names must match the number of row_data.

Reply:
Done, but its actually:
The number of column names must match the number of values in each MonitoringRow.